### PR TITLE
core: voluntary exit

### PR DESCRIPTION
**Pull request by leolara**
Adds support for voluntary exits via a new `DutyExit`.

category: feature
ticket: #553

### CODE REVIEW

File: core/types.go
Comment (corverroos) [inline core/types.go:39]: I would propose we rename this to `DutyExit`, since there is only this type of exit duty, there are no non-voluntary or any other type of exit duty. This also aligns with the other duties above, which is just `DutyProposer` and `DutyRandao` instead of more verbose redundant versions.

File: core/bcast/bcast.go
Comment (corverroos) [inline core/bcast/bcast.go:104]: Please follow above pattern of extracting this to `core.DecodeExitAggSignedData`

File: core/bcast/bcast.go
Comment (corverroos) [inline core/bcast/bcast.go:111]: please use snake case for json logging fields (see guidelines)

Comment (corverroos) [reply core/bcast/bcast.go:111]: I would also add the pubkey here: `z.Any("pubkey", pubkey)`

File: core/bcast/bcast_test.go
Comment (corverroos) [inline core/bcast/bcast_test.go:102]: you could extract this as `testutil.RandomSignedVoluntaryExit`

File: core/bcast/bcast_test.go
Comment (corverroos) [inline core/bcast/bcast_test.go:110]: extract this to `core.EncodeExitAggSignedData`

File: core/sigagg/sigagg.go
Comment (corverroos) [inline core/sigagg/sigagg.go:152]: same here, use `core.Encode/Decode`

File: core/sigagg/sigagg.go
Comment (corverroos) [inline core/sigagg/sigagg.go:203]: same here

File: core/validatorapi/router.go
Comment (corverroos) [inline core/validatorapi/router.go:75]: not sure this is required, why not just support any method since it doesn't really matter?

File: core/validatorapi/validatorapi.go
Comment (corverroos) [inline core/validatorapi/validatorapi.go:370]: snake case please

File: core/validatorapi/validatorapi.go
Comment (corverroos) [inline core/validatorapi/validatorapi.go:373]: this will panic if the map contains a different index.

Comment (corverroos) [reply core/validatorapi/validatorapi.go:373]: can rename variable to `eth2Pubkey`

File: core/validatorapi/validatorapi.go
Comment (corverroos) [inline core/validatorapi/validatorapi.go:406]: can use `core.Encode...`

File: core/validatorapi/validatorapi.go
Comment (corverroos) [inline core/validatorapi/validatorapi.go:411]: I would use slot 0, since a zero value in golang often means "unset" and slot 0 isn't a valid eth2 slot.

File: core/validatorapi/validatorapi.go
Comment (corverroos) [inline core/validatorapi/validatorapi.go:369]: can do `else if` since it is part of the error handling/validator of the above call.

**Comment by corverroos** [comment]
Not sure the description of the PR is correct. This doesn't have anything to do with offline validators? This is just adding support for voluntary exit data...?

File: core/validatorapi/validatorapi.go
Comment (corverroos) [reply core/validatorapi/validatorapi.go:411]: or actually, we should probably convert the epoch into a slot (1 slot in the epoch) and use that. Just incase multiple exits are requested with different epochs.

File: core/bcast/bcast_test.go
Comment (dB2510) [reply core/bcast/bcast_test.go:102]: i think it is too small to extract plus it would be used only here

File: core/validatorapi/router.go
Comment (dB2510) [reply core/validatorapi/router.go:75]: yeah, we don't need methods since we use eth2-http package for handler which internally handles methods and stuff.

File: core/bcast/bcast.go
Comment (leolara) [reply core/bcast/bcast.go]: Done

File: core/validatorapi/validatorapi.go
Comment (leolara) [reply core/validatorapi/validatorapi.go]: Done both things

File: eth2util/signing/signing.go
Comment (leolara) [reply eth2util/signing/signing.go:36]: Done

File: core/validatorapi/validatorapi.go
Comment (leolara) [reply core/validatorapi/validatorapi.go]: I have changed this code, so this `if len' is not anymore

File: core/validatorapi/validatorapi.go
Comment (leolara) [reply core/validatorapi/validatorapi.go:412]: ok, I do it but I will put a reference that it was your idea

Comment (leolara) [reply core/validatorapi/validatorapi.go]: Done

File: core/bcast/bcast_test.go
Comment (leolara) [reply core/bcast/bcast_test.go]: Done

File: core/bcast/bcast.go
Comment (corverroos) [inline core/bcast/bcast.go:80]: can remove `.String()` since `z.Any` does that automatically

File: core/validatorapi/validatorapi.go
Comment (corverroos) [inline core/validatorapi/validatorapi.go:411]: This comment doesn't seem very professional

File: core/validatorapi/validatorapi.go
Comment (corverroos) [inline core/validatorapi/validatorapi.go:412]: I meant converting the epoch into the first slot of the epoch. not just using the epoch itself.

File: testutil/random.go
Comment (corverroos) [inline testutil/random.go:389]: `t` not required since not used

File: core/encode.go
Comment (corverroos) [inline core/encode.go:107]: please remove unprofessional comments

File: core/bcast/bcast.go
Comment (leolara) [reply core/bcast/bcast.go:109]: Done

File: core/bcast/bcast_test.go
Comment (leolara) [reply core/bcast/bcast_test.go:104]: Done

File: core/validatorapi/validatorapi.go
Comment (leolara) [reply core/validatorapi/validatorapi.go]: Done

**Review by corverroos — CHANGES_REQUESTED** [review]
I disagree with putting personal style disagreements into the code as comments.

File: core/validatorapi/validatorapi.go
Comment (leolara) [reply core/validatorapi/validatorapi.go:406]: I am happy to remove it, and you add the code in a commit under your name.

If this is a bad decision, that you or the company should be ashamed of, please let me know why we are taking it, when there are better alternatives.

File: core/validatorapi/router.go
Comment (leolara) [reply core/validatorapi/router.go:75]: We only want to capture the POST, the GET should be proxied

File: testutil/random.go
Comment (leolara) [reply testutil/random.go:380]: t is used in t.Helper like in the other Random things

File: core/sigagg/sigagg.go
Comment (leolara) [reply core/sigagg/sigagg.go:151]: Done

File: core/sigagg/sigagg.go
Comment (leolara) [reply core/sigagg/sigagg.go:193]: Done

File: core/validatorapi/validatorapi.go
Comment (leolara) [reply core/validatorapi/validatorapi.go:407]: that is not what you said, do you want to change it again?

File: core/validatorapi/validatorapi.go
Comment (leolara) [reply core/validatorapi/validatorapi.go:406]: so, do you want me to remove it?

File: core/encode.go
Comment (leolara) [reply core/encode.go:107]: What is unprofessional?

File: core/validatorapi/validatorapi.go
Comment (corverroos) [reply core/validatorapi/validatorapi.go:407]: yes

File: testutil/random.go
Comment (corverroos) [reply testutil/random.go:380]: it isn't used, so you do not need to include it, some random things require it, others don't.

File: core/encode.go
Comment (corverroos) [reply core/encode.go:107]: Putting personal style disagreements into the code as comments

File: testutil/random.go
Comment (leolara) [reply testutil/random.go]: Done

File: testutil/random.go
Comment (leolara) [reply testutil/random.go]: Done

File: testutil/random.go
Comment (leolara) [reply testutil/random.go]: Done

File: core/validatorapi/validatorapi.go
Comment (leolara) [reply core/validatorapi/validatorapi.go:407]: ok so I revert it

File: core/validatorapi/validatorapi.go
Comment (leolara) [reply core/validatorapi/validatorapi.go]: Done

File: core/bcast/bcast.go
Comment (leolara) [reply core/bcast/bcast.go:80]: Done

**Comment by leolara** [comment]
| Not sure the description of the PR is correct. This doesn't have anything to do with offline validators? This is just adding support for voluntary exit data...?

The PR description is from the ticket from @OisinKyne 

if you have a better description please change it.

**Comment by leolara** [comment]
@corverroos if you want to put something else in the Slot of the duty, please push it as a commit to this branch, I am confused about what you want there

File: core/encode.go
Comment (leolara) [reply core/encode.go:107]: This is not a personal style, this is a statement of fact, that other contributors should be aware of.

If you think this decision is shameful, we should decide about why we are doing it this way, when there are easier ways.

Comment (leolara) [reply core/encode.go:107]: You can find in Open Source software this type of comments very often, that using a method is dangerous for example.

Comment (leolara) [reply core/encode.go:107]: https://github.com/ethereum/go-ethereum/blob/07508ac0e9695df347b9dd00d418c25151fbb213/node/config.go#L176

Comment (leolara) [reply core/encode.go:107]: https://github.com/ethereum/go-ethereum/blob/d8ff53dfb8a516f47db37dbc7fd7ad18a1e8a125/accounts/scwallet/README.md?plain=1#L11

Comment (leolara) [reply core/encode.go:107]: https://github.com/ethereum/go-ethereum/blob/8a134014b4b370b4a3632e32a2fc8e84ee2b6947/crypto/bls12381/arithmetic_fallback.go#L26-L30

Comment (leolara) [reply core/encode.go:107]: https://github.com/ethereum/go-ethereum/blob/d8ff53dfb8a516f47db37dbc7fd7ad18a1e8a125/metrics/meter.go#L104

Comment (leolara) [reply core/encode.go:107]: https://github.com/ethereum/go-ethereum/blob/41671d449f5b3a19119eef070151ec72fdbb31f8/build/nsis.envvarupdate.nsh#L5-L6

Comment (leolara) [reply core/encode.go:107]: https://github.com/ethereum/go-ethereum/blob/1941c5e6c99689be2e27b3e537bd591688b717de/core/rawdb/chain_freezer.go#L46-L48

File: core/types.go
Comment (leolara) [reply core/types.go:39]: I don't fell comfortable with this, there will be other Exits in the code, could we ask for feedback to @OisinKyne @dB2510 @xenowits ?

**Comment by leolara** [comment]
@corverroos I have remove the WARNING comment and changed VoluntaryExit to Exit.

The epoch thing I don't understand please add a commit to the branch before merging.

Cheers!

**Review by corverroos — APPROVED** [review]

**Pull request merged**
